File: PowerPathFinder/ppf/common/models/route.py
"""
Data models definition for all PowerPathFinder services, this module serves as single source of
truth. Other services will introspect the DB in order to be up to date. It includes the following
models:

- Route: Describes the routes that drivers can create and passengers can join.
- RoutePassengers: Describes the passengers that are part of a route, records are created when
    users join to a route.
"""

from datetime import timedelta
from django.db import models
from django.db.models import Q

from .user import Driver, User


class RouteManager(models.Manager):
    """
    RouteManager is a custom Manager that adds 'table-level' functionality
    This managers will act as the entry point of the domain to the persistance layer
    https://docs.djangoproject.com/en/5.0/topics/db/managers/#custom-managers
    # See QuerySet API https://docs.djangoproject.com/en/5.0/ref/models/querysets/#queryset-api
    # See Filtered Relations https://docs.djangoproject.com/en/5.0/ref/models/querysets/#queryset-api
    # See Complex Lookups https://docs.djangoproject.com/en/5.0/topics/db/queries/#complex-lookups-with-q-objects
    """

    def byUser(self, userId):
        """
        returns all routes where the user is present either as driver or passenger
        """
        return self.get(Q(driver_id=userId) | Q(passengers__id__contains=userId))

    def active(self):
        """
        active queryset, retrieves all routes that ar not either cancelled or finalized
        """
        return self.filter(cancelled=False, finalized=False)

    def cancelled(self):
        """
        cancelled queryset, retrieves all routes that are cancelled
        """
        return self.filter(cancelled=True)

    def finalized(self):
        """
        finalized queryset, retrieves all routes that are finalized
        """
        return self.filter(finalized=True)

    def withSeats(self):
        """
        withSeats queryset, retrieves all routes that has at least one free seat
        """
        return self.filter(freeSeats__gt=0)


class Route(models.Model):
    """
    Route between two points, organized by a driver to be shared with passengers.
    """

    objects: RouteManager = RouteManager()  # Must assign the custom manager and specify the type

    driver = models.ForeignKey(Driver, on_delete=models.CASCADE)

    originLat = models.FloatField()
    originLon = models.FloatField()
    originAlias = models.CharField(max_length=100)

    destinationLat = models.FloatField()
    destinationLon = models.FloatField()
    destinationAlias = models.CharField(max_length=100)

    polyline = models.TextField()
    waypoints = models.JSONField(default=list)
    distance = models.PositiveIntegerField()
    duration = models.PositiveIntegerField()

    departureTime = models.DateTimeField()
    freeSeats = models.PositiveSmallIntegerField()
    price = models.FloatField(default=0.0)

    passengers = models.ManyToManyField(User, related_name="joined_routes")

    cancelled = models.BooleanField(default=False)
    finalized = models.BooleanField(default=False)

    createdAt = models.DateTimeField(auto_now_add=True)

    class Meta:
        app_label = "common"

    def isFull(self):
        """
        Returns True if the route is full, False otherwise.
        """
        return self.freeSeats == 0

    def overlapsWith(self, routeId):  # TODO refactor to accept Route instance
        """
        Returns True if the route temporally overlaps with the route with the provided ID, False otherwise.
        """
        duration = timedelta(seconds=self.duration)
        route = Route.objects.get(id=routeId)
        if self.departureTime + duration >= route.departureTime:
            return True
        return False

File: ppf_mobile_client/lib/Models/Route.dart
// ignore_for_file: file_names

class MapRoute {
  int id;
  double originLat;
  double originLon;
  String originAlias;
  double destinationLat;
  double destinationLon;
  String destinationAlias;
  String polyline;
  double distance;
  double duration;
  DateTime departureTime;
  int freeSeats;
  double price;
  bool cancelled;
  bool finalized;
  DateTime createdAt;
  int driverId;
  String driverName;
  List<dynamic> waypoints;
  List<int>? passengers;

  MapRoute({
    required this.id,
    required this.originLat,
    required this.originLon,
    required this.originAlias,
    required this.destinationLat,
    required this.destinationLon,
    required this.destinationAlias,
    required this.polyline,
    required this.distance,
    required this.duration,
    required this.departureTime,
    required this.freeSeats,
    required this.price,
    required this.cancelled,
    required this.finalized,
    required this.createdAt,
    required this.driverId,
    required this.driverName,
    required this.waypoints,
    required this.passengers,
  });

  factory MapRoute.fromJson(Map<String, dynamic> json) {
    var driverJson = json['driver'];
    List<dynamic> passengersJson = json['passengers'] as List<dynamic>? ?? [];
    return MapRoute(
      id: json['id'] as int,
      originLat: json['originLat'] as double,
      originLon: json['originLon'] as double,
      originAlias: json['originAlias'] as String,
      destinationLat: json['destinationLat'] as double,
      destinationLon: json['destinationLon'] as double,
      destinationAlias: json['destinationAlias'] as String,
      polyline: json['polyline'] as String,
      distance: (json['distance'] is int)
          ? json['distance'].toDouble()
          : json['distance'] as double,
      duration: (json['duration'] is int)
          ? json['duration'].toDouble()
          : json['duration'] as double,
      departureTime: DateTime.parse(json['departureTime'] as String).toLocal(),
      freeSeats: json['freeSeats'] as int,
      price: json['price'] as double,
      cancelled: json['cancelled'] as bool,
      finalized: json['finalized'] as bool,
      createdAt: DateTime.parse(json['createdAt'] as String).toLocal(),
      driverId: driverJson['id'] as int,
      driverName: driverJson['username'] as String,
      waypoints: json['waypoints'] as List<dynamic>,
      passengers: passengersJson.map((e) => e["id"] as int).toList(),
    );
  }

  // Empty constructor
  MapRoute.empty()
      : id = 0,
        originLat = 0,
        originLon = 0,
        originAlias = '',
        destinationLat = 0,
        destinationLon = 0,
        destinationAlias = '',
        polyline = '',
        distance = 0,
        duration = 0,
        departureTime = DateTime.now(),
        freeSeats = 0,
        price = 0,
        cancelled = false,
        finalized = false,
        createdAt = DateTime.now(),
        driverId = 0,
        driverName = '',
        passengers = [],
        waypoints = [];
}
